File: java/src/main/java/com/tormentnexus/aider/EditBlock.java
package com.tormentnexus.aider;

// Re-implementation of Aider's SEARCH/REPLACE diff format patch application

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class EditBlock {
    public String search;
    public String replace;

    public EditBlock(String search, String replace) {
        this.search = search;
        this.replace = replace;
    }

    public static List<EditBlock> parseEditBlocks(String llmResponse) {
        List<EditBlock> blocks = new ArrayList<>();
        Pattern pattern = Pattern.compile("<<<<<<< SEARCH\n([\\s\\S]*?)=======\n([\\s\\S]*?)>>>>>>> REPLACE");
        Matcher matcher = pattern.matcher(llmResponse);

        while (matcher.find()) {
            blocks.add(new EditBlock(matcher.group(1), matcher.group(2)));
        }
        return blocks;
    }

    public static String applyEditBlocks(String fileContent, List<EditBlock> blocks) {
        String newContent = fileContent;
        for (EditBlock block : blocks) {
            if (newContent.contains(block.search)) {
                newContent = newContent.replace(block.search, block.replace);
            } else {
                System.out.println("Warning: Could not find SEARCH block exactly.");
            }
        }
        return newContent;
    }
}

File: rust/src/aider/edit_block.rs
// Re-implementation of Aider's SEARCH/REPLACE diff format patch application

use regex::Regex;

pub struct EditBlock {
    pub search: String,
    pub replace: String,
}

pub fn parse_edit_blocks(llm_response: &str) -> Vec<EditBlock> {
    let mut blocks = Vec::new();
    let re = Regex::new(r"<<<<<<< SEARCH\n([\s\S]*?)=======\n([\s\S]*?)>>>>>>> REPLACE").unwrap();

    for cap in re.captures_iter(llm_response) {
        blocks.push(EditBlock {
            search: cap[1].to_string(),
            replace: cap[2].to_string(),
        });
    }
    blocks
}

pub fn apply_edit_blocks(file_content: &str, blocks: &[EditBlock]) -> String {
    let mut new_content = file_content.to_string();
    for block in blocks {
        if new_content.contains(&block.search) {
            new_content = new_content.replace(&block.search, &block.replace);
        } else {
            println!("Warning: Could not find SEARCH block exactly.");
        }
    }
    new_content
}
